### ISSUE TITLE
Publish v1-0 version of the docs

### ISSUE THREAD

**Issue by jlewi** [kubeflow/website#1581]
We'd like to do the following

1. Start publishing master on v1-0.kubeflow.org
1. Have www.kubeflow.org continue to direct to the v0.7 version of the docs
1. When the docs on master are ready redirect www.kubeflow.org to the v1 version of the docs.

I think doing the above should be fairly straightforward based on the [docs](https://github.com/kubeflow/kubeflow/blob/master/docs_dev/releasing.md#creating-a-website-branch-for-the-latest-major-or-minor-release)

We basically want to follow the [process](https://github.com/kubeflow/kubeflow/blob/master/docs_dev/releasing.md#creating-a-website-branch-for-the-latest-major-or-minor-release) for cutting release branches.

The only difference is in netlify. We should point master/latest at the version of the site corresponding to the v0.7 branch not v1.0.

**Comment by jlewi** [comment]
In Netlify I used [deploy contexts](https://docs.netlify.com/site-deploys/overview/#deploy-contexts)
to switch the production branch from master to v0-7

I also enabled branch deploys for all branches

**Comment by jlewi** [comment]
I believe the current [process](https://github.com/kubeflow/kubeflow/blob/master/docs_dev/releasing.md#creating-a-website-branch-for-the-latest-major-or-minor-release) is creating a new netlify site for each branch. I'm not sure we want to do that. I think all of those cites are being owned by the individuals who created them; they don't show up in my netlify team when I log in.

**Comment by jlewi** [comment]
I enabled deploy branches but they don't seem to be deployed; maybe we need to push a commit to trigger deploy?

At the very least the production version of the website www.kubeflow.org is now pointing to the v0.7-branch so we should be able to push changes to master.

**Comment by jlewi** [comment]
When I pushed a commit to the v0.5-branch that triggered a deploy but it doesn't appear to be showing up at https://v0.5-branch--competent-brattain-de2d6d.netlify.com/

Could the period in the name be problematic?

**Comment by jlewi** [comment]
Looks like the period is the problem. I pushed a commit to the branch will-may and that branch is now visible at
https://will-may--competent-brattain-de2d6d.netlify.com/

So I think next steps

* It looks like https://master--competent-brattain-de2d6d.netlify.com/ is already available
* I updated the CNAME record for v1-0.kubeflow.org so it should point to master
* I think we need to update our [proccess](https://github.com/kubeflow/kubeflow/blob/master/docs_dev/releasing.md#creating-a-website-branch-for-the-latest-major-or-minor-release)
   * We shouldn't create new sites for new branches
   * Going forward we should avoid "." its the branch names
   * We should document how to update the

**Comment by jlewi** [comment]
I take it back looks like netlify just replaces "." with "-"
https://v0-5-branch--competent-brattain-de2d6d.netlify.com/

**Comment by jlewi** [comment]
It looks like Netlify doesn't create a branch for the site set to production;
i.e.
https://v0-7-branch--competent-brattain-de2d6d.netlify.com/ doesn't exist. That's probably because in the deploy settings I set the production branch to v0.7-branch.

I'm going to change it back to master.

**Comment by jlewi** [comment]
https://v0-7-branch--competent-brattain-de2d6d.netlify.com/ is now live.

I've updated the CNAME records as below.
![cnames](https://user-images.githubusercontent.com/777219/73553948-ae681680-43ff-11ea-86f1-d6e066cd246c.png)

Here are the results of nslookup.

```
jlewi@jlewi-glaptop2:~/git_kubeflow-website$ nslookup www.kubeflow.org
Server:		127.0.0.1
Address:	127.0.0.1#53

Non-authoritative answer:
www.kubeflow.org	canonical name = v0-7-branch--competent-brattain-de2d6d.netlify.com.
Name:	v0-7-branch--competent-brattain-de2d6d.netlify.com
Address: 104.248.78.24
Name:	v0-7-branch--competent-brattain-de2d6d.netlify.com
Address: 2604:a880:2:d1::1a9:8001

jlewi@jlewi-glaptop2:~/git_kubeflow-website$ nslookup v1-0.kubeflow.org
Server:		127.0.0.1
Address:	127.0.0.1#53

Non-authoritative answer:
v1-0.kubeflow.org	canonical name = competent-brattain-de2d6d.netlify.com.
Name:	competent-brattain-de2d6d.netlify.com
Address: 165.227.0.164
Name:	competent-brattain-de2d6d.netlify.com
Address: 2604:a880:2:d1::1a9:8001
```

**Comment by jlewi** [comment]
So we should be good to go to be begin pushing 1.0 changes to master.

**Comment by jlewi** [comment]
Hmm. Looks like https://v1-0.kubeflow.org/ doesn't work. We get an SSL certificate error. We should still be able to access them at 
competent-brattain-de2d6d.netlify.com

In Netlify I added a domain alias for "v1-0.kubeflow.org" That worked; but I think it might be pointing to the v0.7 branch now.

**Comment by jlewi** [comment]
Looks like master might still be showing up on www.kubeflow.org. I merged #1520 which should only show up master.

The v0-7.kubeflow.org appears to be correctly pointing the v0.7 branch

The following URLs return 404 since #1520 is not on the branch
https://v0-7.kubeflow.org/docs/other-guides/support/
https://v0-7-branch--competent-brattain-de2d6d.netlify.com/docs/other-guides/support/

https://v0-7.kubeflow.org/
however works

It is also available on the production site
https://competent-brattain-de2d6d.netlify.com/docs/other-guides/support/
which is also correct as that should point to master
https://master--competent-brattain-de2d6d.netlify.com/docs/other-guides/support/
also works.

But it is also available at  
https://www.kubeflow.org/docs/other-guides/support/
which indicates www.kubeflow.org is incorrectly pointing at master and not the v0.7 branch

**Comment by jlewi** [comment]
I suspect we may need to enable branch subdomains with Netlify.

**Comment by jlewi** [comment]
Here's the original issue when we setup Netlify
https://github.com/kubeflow/website/issues/5#issuecomment-389037774

* I don't think we are actually using Netlify DNS; I think our nameservers are still pointing at googedomains and clouddns
* In Netlify I changed the Deploy settings to make v0.7-branch the production branch to see if we can use that to redirect www.kubeflow.org to 0.7

**Comment by jlewi** [comment]
I triggered a redeploy; 
www.kubeflow.org now appears to be pushing to v0.7-branch because
https://www.kubeflow.org/docs/other-guides/support/ - is no longer accessible

**Comment by jlewi** [comment]
https://master.kubeflow.org/docs/other-guides/support/ is not accessible so it doesn't appear to be pointing to the master branch.

In Netlify it appears to be configured as a domain alias.

https://v0-6.kubeflow.org/docs/gke/deploy/deploy-cli/
This is correctly pointing to the v0.6 branch; the header at the top shows its the archive version.

**Comment by jlewi** [comment]
Lets try changing the cname record in Cloud DNS to point to 
master--competent-brattain-de2d6d.netlify.com.

**Comment by jlewi** [comment]
```
curl https://v1-0.kubeflow.org
curl: (60) SSL: no alternative certificate subject name matches target host name 'v1-0.kubeflow.org'
More details here: https://curl.haxx.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above
```

So right now

* www.kubeflow.org is correctly mapped to v0.7-branch
* https://master--competent-brattain-de2d6d.netlify.com/docs/other-guides/support/ is redirecting to master branch

This is workable for now.

* It looks like the problem with setting up CNAME records for subdomains is the Netlify doesn't know about them and generate certs for them.
  * I suspect the reason the other subdomains are working e.g "v0-5.kubeflow.org" is because in the past we set up separate sites for those domains and that's whats being used.

**Comment by jlewi** [comment]
I think what we have is good enough for now.

To summarize

1. In Netlify we use Deploy contexts to switch the branch production is pointing to
1. We can access a branch using `http://${BRANCH}---competent-brattain-de2d6d.netlify.com`
1. TODO: We still need to figure out how to map a subdomain like ${BRANCH}.kubeflow.org to a branch like `http://${BRANCH}---competent-brattain-de2d6d.netlify.com`
     * References [Using Deploy Branches with external DNS](https://community.netlify.com/t/common-issue-how-to-use-netlify-s-branch-deploy-feature-without-netlify-dns/128)
     * I tried this in Netlify domains I created a branch subdomain v0.7-branch which points at v0-7-branch.kubeflow.org
     * I clicked renew certificate to generate a certificate for it
     * I created a CNAME record for it in Cloud DNS
     * The problem is v0-7-branch.kubeflow.org is redirecting to kubeflow.org
     * Lets give it an hour to see if its just a DNS propogation issue.

**Comment by jlewi** [comment]
I tried removing master.kubeflow.org as an alias in Netlify. We'd like to use branch subdomains for that.

**Comment by jlewi** [comment]
master is now listed in branch subdomains. 
When I try to access: https://master.kubeflow.org/
I get an HTTPs error. I think we may need to wait for the DNS record to propogate and then in Netlify renew the certificate.

Right now in Netlify under https it doesn't show a certificate for master.kubeflow.org

**Comment by jlewi** [comment]
It looks like the DNS entry is set but the certificate still isn't valid

If I do the following I get the page as expected

```
curl --insecure https://master.kubeflow.org/docs/other-guides/support/
```

Could this be the issue mentioned in the [docs](https://community.netlify.com/t/common-issue-how-to-use-netlify-s-branch-deploy-feature-without-netlify-dns/128) about SSL support and external DNS?

Why is the v0-7 subdomain still redirecting  to

```
curl --insecure https://v0-7-branch.kubeflow.org/docs/other-guides/support/
Redirecting to https://www.kubeflow.org/docs/other-guides/support/
```

**Comment by sarahmaddox** [comment]
@jlewi The website at www.kubeflow.org now declares itself to be an unmaintained version of the docs, and gives people a link for up to date docs which points them back to the same place:

![image](https://user-images.githubusercontent.com/6138251/73584248-f2810880-444b-11ea-8250-335b53f50805.png)

**Comment by sarahmaddox** [comment]
My recommendation is to keep www.kubeflow.org reflecting the master, without needing to do any redirects. When we're ready to start publishing the docs for an RC (in this case, the v1.0.0 RCs) then we should (a) make sure we've already created the archive version for the current docs (in this case v07. which we have created) and (b) change the version number on the master and on www.kubeflow.org to represent the new version (in this case, **v1.0 RC**) and add a banner on every page saying the docs now reflect the upcoming release. We can link people back to the latest archive (in this case, v0.7) for the stable release.

**Comment by jlewi** [comment]
@sarahmaddox Can we just remove the banner on the v0.7-branch?

When people go to www.kubeflow.org they should see the v0.7 version of the docs.
I would rather not direct users to 1.0 docs and then force them to redirect. If we show 1.0 docs on www.kubeflow.org that is an invitation for people to try it out and get frustrated when things don't work.

**Comment by jlewi** [comment]
The master version of the docs are now available on
https://master.kubeflow.org/
But you will get a 

`NET::ERR_CERT_COMMON_NAME_INVALID

This server could not prove that it is master.kubeflow.org; its security certificate is from *.netlify.com. This may be caused by a misconfiguration or an attacker intercepting your connection.
`

In chrome you should be able to click on advanced and then just proceed. It looks like you can't be signed to chrome using your corporate account; using icognito or your personal account should work.

**Comment by jlewi** [comment]
v0-7-branch.kubeflow.org still redirects to master.kubeflow.org I wonder if this is because we've configured kubeflow.org to direct to the the v0.7 branch?

kubeflow/website#1586 for the cert error.

**Comment by sarahmaddox** [comment]
I cannot access master.kubeflow.org on either my corp or my personal account in Chrome. I can access it in incognito.

Having the site show old, unmaintained docs is bad UX for contributors, whether we show the banner or not - they'll see a version of the docs that doesn't reflect what's in the repo. When their pull requests are merged, they won't see the update reflected in the docs. IMO we therefore need a banner of some sort, if we must show people the v0.7 docs.

If we're ready to start publishing the docs for the RC, then don't we want people to try it (in the full knowledge that it's an early release)?

I think it's confusing to have the site redirecting to an older branch for part of its life - this could last a couple of weeks for v1.0.

@abhi-g and @theadactyl I'm keen to know your opinion here. We did discuss this back in August 2018 (I can forward you the notes) and decided that we should publish master even if experimental. At that time, this was the policy followed by the other projects we looked at. I'm happy to discuss it again, but I do want to make sure we all agree.

**Comment by jlewi** [comment]
@sarahmaddox what about https://master--competent-brattain-de2d6d.netlify.com/?

>Having the site show old, unmaintained docs is bad UX for contributors, whether we show the banner or not - they'll see a version of the docs that doesn't reflect what's in the repo.

I agree its bad for contributors but what about users? If we start publishing 1.0 docs at www.kubeflow.org and showing them to users isn't that exposing users to our experiments?

My concern is that users will go to www.kubeflow.org try out 1.0, complain about it, and then we will tell them its not ready and to try v0.7.

How do we get the docs ready for 1.0 without exposing existing users to the churn and inconsistencies? Can we have some way to atomically publish all the 1.0 changes at once?

**Comment by sarahmaddox** [comment]
Yes, I can access https://master--competent-brattain-de2d6d.netlify.com/.

This also now works: https://v0-7.kubeflow.org/

I don't have a way to automatically merge all the PRs when we're ready, nor to atomically publish all the v1.0 docs at once. I still favor the solution of showing the experimental version on the main doc site when we have a reasonably stable RC, and publishing the docs serially from that point onwards. 

I do see what you're saying about users, but whichever way we go, we're giving users a bad experience. If we redirect to a snapshot, users don't know that they're not getting the latest docs. They won't see the latest version of the docs unless we cherry-pick bug fixes. This could last a couple of weeks (or more) of each quarter.

However, I'm happy to go with the new solution, given that we seem to have opposing opinions here. To make things work, I think this is what needs to happen:

* Make master.kubeflow.org accessible - fix the certificate error. This should be a P0 as the situation is currently not workable.

* Update the [release procedures](https://github.com/kubeflow/kubeflow/blob/master/docs_dev/releasing.md), for at least the following: 

   * Add the Netlify procedure for setting and resetting the redirect.
   * Make any other necessary changes to the Netlify procedures to reflect the updates made in this issue.
   * In the doc versioning procedure, don't enable the "archived" banner until the redirect is reset.

* Update the docs README to let people know that their changes may not be immediately visible on the website, because of the redirects around release time, but that they can see their change in master.

* Check all release procedures and docs procedures for any other necessary updates.

**Comment by jlewi** [comment]
@sarahmaddox I think we can separate this into two questions

1. What do we for PRs that are aimed at the upcoming release
1. What do we show to users

We can either point www.kubeflow.org at the v0.7-branch or at master.

The reason I think we should point www.kubeflow.org at the v0.7-branch is that this unblocks 1.0 development on master.

The alternative would be to treat master as "v0.7" and do development of 1.0 on a branch. 

I believe our reasoning in the past (and consistent for non-doc development) was to create stable release branches for users and then continue development on master.

If we think we should show users the latest docs on master even though they are under development we can repoint www.kubeflow.org at master.

The primary thing I want to address is unblocking PRs related to 1.0. e.g.

#1583
#1520 
etc...

Merging these PRs into a branch seems preferable to putting holds on them until we are ready to review them and then trying to integrate them all at once. It also makes it easier to get a consistent view of what state the docs are in for 1.0.

I don't think we want to start merging 1.0 PRs though if they are directly visible to users. That's why i think we should point www.kubeflow.org at the v0.7 branch.

Once we are ready to make 1.0 docs visible we would point www.kubeflow.org back at master. 

So master and www.kubeflow.org would only be out of sync for the short period leading up to a release during which we are trying to get the docs ready for 1.0 but not yet ready to publish them.

**Comment by jlewi** [comment]
For the SSL certificate issue; the [docs](https://community.netlify.com/t/common-issue-how-to-use-netlify-s-branch-deploy-feature-without-netlify-dns/128) say to contact technical support. So I submitted a request.

**Comment by jlewi** [comment]
Certificate has been fixed by netlfiy support.

I was able to access https://master.kubeflow.org/  with no more warnings.

kubeflow/kubeflow#4717 is pending to update instructions.

**Comment by abhi-g** [comment]
Awesome.

I think for Kubeflow Users the current release version is 0.7.x until we
release 1.0. Therefore www.kubeflow.org should be pointing to stable docs
from 0.7.

From developer perspective:
 - 0.7 docs are mostly stable and shouldn't need changes. If in off cases
they do, then the developers can merge cherry picks, or make modifications
in 0.7 branch.
- they should have easy way to update documents for upcoming release and
therefore should merge their update PRs to master.

So based on this, we have three categories for docs:

1. Archival: accessible to users via archival sites from versioned
branches. Developers can update these by merged to appropriate branch, an
uncommon activity.
2. Current: last release that is recommended for deployment to users. This
set of docs are stable and accessible to users via the site for
corresponding branch. But since this is the most common set of documents
that users should be accessing, this is what www.kubeflow.org should point
to. Updates to these from devs should again be less frequent and therefore
requiring cherry picks or merge into the branch corresponding to current
release is reasonable.
3. Upcoming: Most users should not be working with set until release is
final. Users accessing these should understand risks and be okay to use the
bleeding edge. Developers will most often be updating these documents as
they get features ready. Therefore master.kubeflow.org is an appropriate
site to access these. Also, for devs this matches seamlessly with their
development workflow into master.

@sarahmaddox @jlewi Hopefully the above reconciles our issues. Sarah this
evolves a bit from our previous discussion last year but I think moves us
in the right direction.

**Comment by sarahmaddox** [comment]
Thanks @abhi-g and @jlewi

**Comment by jlewi** [comment]
Downgrading to P1 because I don't think the remaining work is blocking 1.0.
Remaining work is to get kubeflow/kubeflow#4717 merged updating the process for releasing the docs.

**Comment by sarahmaddox** [comment]
@jlewi PR https://github.com/kubeflow/kubeflow/pull/4717 is now merged. I think we can close this issue now.

We'll need to repoint www.kubeflow.org to master at some point, but I think the main goal of this particular issue was to publish the v1.0 docs on master.kubeflow.org, which is now done.

Please do re-open this issue if there's still work to do on it.

/close
